File: src/lib/mpa/computations.ts
/**
 * Cost Computations for Monthly Performance Analysis
 *
 * Computes labor costs, expense costs, and merges into project tables
 */

import type {
  HoursRecord,
  ExpenseRecord,
  CompensationRecord,
  RevenueCenter,
  CostCenter,
  NonRevenueClient,
  HoursDetail,
  ExpenseDetail,
} from './types';

interface LaborSummary {
  contractCode: string;
  hours: number;
  laborCost: number;
}

interface ExpenseSummary {
  contractCode: string;
  expenseCost: number;
}

/**
 * Calculate labor costs by joining hours with compensation rates
 */
export function calculateLaborCosts(
  hoursRecords: HoursRecord[],
  compRecords: CompensationRecord[]
): {
  laborSummary: LaborSummary[];
  hoursDetail: HoursDetail[];
  logs: string[];
} {
  const logs: string[] = [];

  // Create compensation lookup
  const compLookup = new Map<string, number>();
  for (const c of compRecords) {
    compLookup.set(c.staffKey, c.hourlyCost);
  }

  // Join hours with compensation
  const joinedRecords: Array<HoursRecord & { hourlyCost: number; laborCost: number }> = [];
  const missingStaff = new Set<string>();
  let missingHours = 0;

  for (const h of hoursRecords) {
    const hourlyCost = compLookup.get(h.staffKey);

    if (hourlyCost === undefined) {
      missingStaff.add(h.staffKey);
      missingHours += h.hours;
      continue;
    }

    joinedRecords.push({
      ...h,
      hourlyCost,
      laborCost: h.hours * hourlyCost,
    });
  }

  if (missingStaff.size > 0) {
    logs.push(
      `${missingStaff.size} staff missing compensation records (${missingHours.toFixed(1)} hours excluded)`
    );
  }

  // Create hours detail (for drill-down) - aggregate by contract_code + staff_key
  const detailMap = new Map<string, HoursDetail>();
  for (const r of joinedRecords) {
    const key = `${r.contractCode}|${r.staffKey}`;
    const existing = detailMap.get(key);

    if (existing) {
      existing.hours += r.hours;
      existing.laborCost += r.laborCost;
    } else {
      detailMap.set(key, {
        contractCode: r.contractCode,
        staffKey: r.staffKey,
        hours: r.hours,
        hourlyCost: r.hourlyCost,
        laborCost: r.laborCost,
      });
    }
  }

  const hoursDetail = Array.from(detailMap.values());

  // Aggregate by project for summary
  const summaryMap = new Map<string, LaborSummary>();
  for (const r of joinedRecords) {
    const existing = summaryMap.get(r.contractCode);

    if (existing) {
      existing.hours += r.hours;
      existing.laborCost += r.laborCost;
    } else {
      summaryMap.set(r.contractCode, {
        contractCode: r.contractCode,
        hours: r.hours,
        laborCost: r.laborCost,
      });
    }
  }

  const laborSummary = Array.from(summaryMap.values());

  return { laborSummary, hoursDetail, logs };
}

/**
 * Calculate expense costs (already filtered to non-reimbursable)
 */
export function calculateExpenseCosts(
  expenseRecords: ExpenseRecord[]
): {
  expenseSummary: ExpenseSummary[];
  expenseDetail: ExpenseDetail[];
} {
  // Keep detail for drill-down
  const expenseDetail: ExpenseDetail[] = expenseRecords.map(e => ({
    contractCode: e.contractCode,
    expenseDate: e.date,
    amount: e.amount,
    notes: e.notes,
  }));

  // Aggregate by project
  const summaryMap = new Map<string, number>();
  for (const e of expenseRecords) {
    const existing = summaryMap.get(e.contractCode) || 0;
    summaryMap.set(e.contractCode, existing + e.amount);
  }

  const expenseSummary: ExpenseSummary[] = Array.from(summaryMap.entries()).map(
    ([contractCode, expenseCost]) => ({ contractCode, expenseCost })
  );

  return { expenseSummary, expenseDetail };
}

/**
 * Merge direct costs into revenue centers table
 */
export function mergeDirectCostsToRevenueCenters(
  revenueCenters: RevenueCenter[],
  laborSummary: LaborSummary[],
  expenseSummary: ExpenseSummary[]
): RevenueCenter[] {
  const laborLookup = new Map<string, LaborSummary>();
  for (const l of laborSummary) {
    laborLookup.set(l.contractCode, l);
  }

  const expenseLookup = new Map<string, ExpenseSummary>();
  for (const e of expenseSummary) {
    expenseLookup.set(e.contractCode, e);
  }

  return revenueCenters.map(rc => {
    const labor = laborLookup.get(rc.contractCode);
    const expense = expenseLookup.get(rc.contractCode);

    return {
      ...rc,
      hours: labor?.hours ?? 0,
      laborCost: labor?.laborCost ?? 0,
      expenseCost: expense?.expenseCost ?? 0,
    };
  });
}

/**
 * Calculate costs for cost centers from hours and expenses
 */
export function calculateCostCenterCosts(
  costCenters: CostCenter[],
  hoursDetail: HoursDetail[],
  expenseDetail: ExpenseDetail[]
): CostCenter[] {
  const ccCodes = new Set(costCenters.map(cc => cc.contractCode));

  // Aggregate labor by cost center
  const laborByCC = new Map<string, { hours: number; laborCost: number }>();
  for (const h of hoursDetail) {
    if (!ccCodes.has(h.contractCode)) continue;

    const existing = laborByCC.get(h.contractCode) || { hours: 0, laborCost: 0 };
    existing.hours += h.hours;
    existing.laborCost += h.laborCost;
    laborByCC.set(h.contractCode, existing);
  }

  // Aggregate expenses by cost center
  const expensesByCC = new Map<string, number>();
  for (const e of expenseDetail) {
    if (!ccCodes.has(e.contractCode)) continue;

    const existing = expensesByCC.get(e.contractCode) || 0;
    expensesByCC.set(e.contractCode, existing + e.amount);
  }

  return costCenters.map(cc => {
    const labor = laborByCC.get(cc.contractCode);
    const expenses = expensesByCC.get(cc.contractCode) || 0;

    return {
      ...cc,
      hours: labor?.hours ?? 0,
      laborCost: labor?.laborCost ?? 0,
      expenseCost: expenses,
      totalCost: (labor?.laborCost ?? 0) + expenses,
    };
  });
}

/**
 * Calculate costs for non-revenue clients from hours and expenses
 */
export function calculateNonRevenueClientCosts(
  nonRevenueClients: NonRevenueClient[],
  hoursDetail: HoursDetail[],
  expenseDetail: ExpenseDetail[]
): NonRevenueClient[] {
  if (nonRevenueClients.length === 0) {
    return nonRevenueClients;
  }

  const nrcCodes = new Set(nonRevenueClients.map(nrc => nrc.contractCode));

  // Aggregate labor
  const laborByNRC = new Map<string, { hours: number; laborCost: number }>();
  for (const h of hoursDetail) {
    if (!nrcCodes.has(h.contractCode)) continue;

    const existing = laborByNRC.get(h.contractCode) || { hours: 0, laborCost: 0 };
    existing.hours += h.hours;
    existing.laborCost += h.laborCost;
    laborByNRC.set(h.contractCode, existing);
  }

  // Aggregate expenses
  const expensesByNRC = new Map<string, number>();
  for (const e of expenseDetail) {
    if (!nrcCodes.has(e.contractCode)) continue;

    const existing = expensesByNRC.get(e.contractCode) || 0;
    expensesByNRC.set(e.contractCode, existing + e.amount);
  }

  return nonRevenueClients.map(nrc => {
    const labor = laborByNRC.get(nrc.contractCode);
    const expenses = expensesByNRC.get(nrc.contractCode) || 0;

    return {
      ...nrc,
      hours: labor?.hours ?? 0,
      laborCost: labor?.laborCost ?? 0,
      expenseCost: expenses,
      totalCost: (labor?.laborCost ?? 0) + expenses,
    };
  });
}

File: api/py/mpa/lib/computations.py
"""
Cost Computations for Monthly Performance Analysis

Computes:
1. Labor costs (hours x hourly_cost)
2. Expense costs (non-reimbursable only)
3. Direct cost merging into revenue table
"""

import pandas as pd
from typing import List, Tuple


def calculate_labor_costs(
    hours_df: pd.DataFrame,
    comp_df: pd.DataFrame
) -> Tuple[pd.DataFrame, pd.DataFrame, List[str]]:
    """
    Calculate labor costs by joining hours with compensation rates.

    Args:
        hours_df: Harvest hours data (contract_code, staff_key, hours)
        comp_df: Compensation data (staff_key, hourly_cost)

    Returns:
        Tuple of:
        - DataFrame with labor costs by contract_code
        - DataFrame with detailed hours (for drill-down)
        - List of log messages

    Note:
        Staff missing from compensation file will be excluded from labor cost calculations
        and flagged as a warning.
    """
    logs = []

    merged = hours_df.merge(
        comp_df[["staff_key", "hourly_cost"]],
        on="staff_key",
        how="left",
    )

    missing = merged[merged["hourly_cost"].isna()]
    if len(missing) > 0:
        missing_staff = sorted(set(missing["staff_key"].astype(str)))
        missing_hours = missing["hours"].sum()
        logs.append(f"{len(missing_staff)} staff missing compensation records ({missing_hours:.1f} hours excluded)")
        merged = merged[merged["hourly_cost"].notna()]

    merged["labor_cost"] = merged["hours"] * merged["hourly_cost"]

    # Create detailed hours for drill-down (aggregated by contract_code + staff_key)
    hours_detail = (
        merged.groupby(["contract_code", "staff_key"]).agg({
            "hours": "sum",
            "hourly_cost": "first",
            "labor_cost": "sum",
        }).reset_index()
    )

    # Aggregate by project for summary
    labor_by_project = (
        merged.groupby("contract_code").agg({
            "hours": "sum",
            "labor_cost": "sum",
        }).reset_index()
    )

    return labor_by_project, hours_detail, logs


def calculate_expense_costs(expenses_df: pd.DataFrame) -> Tuple[pd.DataFrame, pd.DataFrame]:
    """
    Calculate expense costs (already filtered to non-reimbursable).

    Args:
        expenses_df: Filtered expenses data (contract_code, amount)

    Returns:
        Tuple of:
        - DataFrame with expense costs by contract_code
        - DataFrame with expense detail (for drill-down)
    """
    # Keep detail for drill-down
    expense_detail = expenses_df[['contract_code', 'date', 'amount', 'notes']].copy()
    expense_detail.rename(columns={'date': 'expense_date'}, inplace=True)

    # Aggregate by project
    expense_by_project = (
        expenses_df.groupby("contract_code").agg({"amount": "sum"}).reset_index()
    )
    expense_by_project.rename(columns={"amount": "expense_cost"}, inplace=True)

    return expense_by_project, expense_detail


def merge_direct_costs(
    revenue_df: pd.DataFrame,
    labor_df: pd.DataFrame,
    expense_df: pd.DataFrame
) -> pd.DataFrame:
    """
    Merge direct costs into revenue centers table.

    Args:
        revenue_df: Revenue centers from Pro Forma
        labor_df: Labor costs by contract_code
        expense_df: Expense costs by contract_code

    Returns:
        DataFrame with revenue + direct costs
    """
    out = revenue_df.merge(labor_df, on="contract_code", how="left")
    out = out.merge(expense_df, on="contract_code", how="left")
    out["hours"] = out.get("hours", 0).fillna(0.0)
    out["labor_cost"] = out.get("labor_cost", 0).fillna(0.0)
    out["expense_cost"] = out.get("expense_cost", 0).fillna(0.0)
    return out


def calculate_cost_center_costs(
    cost_centers_df: pd.DataFrame,
    hours_detail_df: pd.DataFrame,
    expense_detail_df: pd.DataFrame
) -> pd.DataFrame:
    """
    Calculate costs for cost centers from hours and expenses.

    Args:
        cost_centers_df: Cost centers with contract_code, description, pool
        hours_detail_df: Detailed hours with labor_cost
        expense_detail_df: Detailed expenses with amount

    Returns:
        Cost centers DataFrame with hours, labor_cost, expense_cost, total_cost
    """
    # Aggregate labor by cost center
    cc_labor = hours_detail_df[
        hours_detail_df['contract_code'].isin(cost_centers_df['contract_code'])
    ].groupby('contract_code').agg({
        'hours': 'sum',
        'labor_cost': 'sum'
    }).reset_index()

    # Aggregate expenses by cost center
    cc_expenses = expense_detail_df[
        expense_detail_df['contract_code'].isin(cost_centers_df['contract_code'])
    ].groupby('contract_code').agg({
        'amount': 'sum'
    }).reset_index()
    cc_expenses.rename(columns={'amount': 'expense_cost'}, inplace=True)

    # Merge into cost centers
    result = cost_centers_df.merge(cc_labor, on='contract_code', how='left')
    result = result.merge(cc_expenses, on='contract_code', how='left')

    # Fill NaN and calculate total
    result['hours'] = result['hours'].fillna(0.0)
    result['labor_cost'] = result['labor_cost'].fillna(0.0)
    result['expense_cost'] = result['expense_cost'].fillna(0.0)
    result['total_cost'] = result['labor_cost'] + result['expense_cost']

    return result


def calculate_non_revenue_client_costs(
    non_revenue_df: pd.DataFrame,
    hours_detail_df: pd.DataFrame,
    expense_detail_df: pd.DataFrame
) -> pd.DataFrame:
    """
    Calculate costs for non-revenue clients from hours and expenses.

    Args:
        non_revenue_df: Non-revenue clients with contract_code
        hours_detail_df: Detailed hours with labor_cost
        expense_detail_df: Detailed expenses with amount

    Returns:
        Non-revenue clients DataFrame with hours, labor_cost, expense_cost, total_cost
    """
    if non_revenue_df.empty:
        return non_revenue_df

    # Aggregate labor
    nrc_labor = hours_detail_df[
        hours_detail_df['contract_code'].isin(non_revenue_df['contract_code'])
    ].groupby('contract_code').agg({
        'hours': 'sum',
        'labor_cost': 'sum'
    }).reset_index()

    # Aggregate expenses
    nrc_expenses = expense_detail_df[
        expense_detail_df['contract_code'].isin(non_revenue_df['contract_code'])
    ].groupby('contract_code').agg({
        'amount': 'sum'
    }).reset_index()
    nrc_expenses.rename(columns={'amount': 'expense_cost'}, inplace=True)

    # Merge
    result = non_revenue_df.merge(nrc_labor, on='contract_code', how='left')
    result = result.merge(nrc_expenses, on='contract_code', how='left')

    # Fill and calculate
    result['hours'] = result['hours'].fillna(0.0)
    result['labor_cost'] = result['labor_cost'].fillna(0.0)
    result['expense_cost'] = result['expense_cost'].fillna(0.0)
    result['total_cost'] = result['labor_cost'] + result['expense_cost']

    return result
